src/image/detail/MaskDict.cc

```diff
@@ -31,18 +32,48 @@ class MaskDict::GlobalState {
 
     using HandleList = std::set<std::weak_ptr<MaskDict>, std::owner_less<std::weak_ptr<MaskDict>>>;
 
-    static GlobalState & get() {
-        static GlobalState instance;
-        return instance;
-    }
+    /*
+     * RAII guard for GlobalState.
+     *
+     * Default-constructing a Lock instance is the only way to access
+     * GlobalState - this acquires GlobalState's mutex, ensuring only one
+     * thread has access to its attributes at one time.  The mutex is released
+     * when the Lock instance goes out of scope.
+     */
+    class Lock {
+    public:
+
+        Lock() :
+            allMaskDicts(_get()._allMaskDicts),
+            defaultMaskDict(_get()._defaultMaskDict),
+            _guard(_get()._mutex)
+        {}
+
+        ~Lock() = default;
+
+        HandleList & allMaskDicts;
+        std::shared_ptr<MaskDict> & defaultMaskDict;
+
+    private:
 
-    HandleList allMaskDicts;  // all the live MaskDicts
-    std::shared_ptr<MaskDict> defaultMaskDict;  // default MaskDict to use
+        Lock(Lock const &) = delete;
+        Lock(Lock &&) = delete;
+
+        Lock & operator=(Lock const &) = delete;
+        Lock & operator=(Lock &&) = delete;
+
+        std::lock_guard<std::mutex> _guard;
+    };
 
 private:
 
-    GlobalState() : defaultMaskDict(new MaskDict(true)) {
-        allMaskDicts.insert(defaultMaskDict);
+    static GlobalState & _get() {
+        static GlobalState instance;
+        return instance;
+    }
```

Comment:
Consider documenting that this method is [thread-safe](https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables), since calls to `_get()` are never guarded by a lock. I believe this is the *only* thread-safe method (counting public field access) in the class.